ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -279,58 +326,52 @@ IoRegisterPlugPlayNotification(IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
-            RtlCopyMemory(&NotificationInfos.InterfaceClassGuid,
-                          EventCategoryData,
-                          sizeof(GUID));
-            NotificationInfos.SymbolicLinkName = &SymbolicLinkU;
-
-            for (SymbolicLink = SymbolicLinkList;
-                 *SymbolicLink;
-                 SymbolicLink += wcslen(SymbolicLink) + 1)
-            {
-                RtlInitUnicodeString(&SymbolicLinkU, SymbolicLink);
-                DPRINT("Calling callback routine for %S\n", SymbolicLink);
-                (*CallbackRoutine)(&NotificationInfos, Context);
-            }
-
-            ExFreePool(SymbolicLinkList);
-        }
-    }
-
     Entry->PnpNotificationProc = CallbackRoutine;
     Entry->EventCategory = EventCategory;
     Entry->Context = Context;
     Entry->DriverObject = DriverObject;
+
     switch (EventCategory)
     {
         case EventCategoryDeviceInterfaceChange:
         {
-            Status = RtlStringFromGUID(EventCategoryData, &Entry->Guid);
-            if (!NT_SUCCESS(Status))
+            if (EventCategoryFlags & PNPNOTIFY_DEVICE_INTERFACE_INCLUDE_EXISTING_INTERFACES)
             {
-                ExFreePoolWithTag(Entry, TAG_PNP_NOTIFY);
-                ObDereferenceObject(DriverObject);
-                return Status;
+                DEVICE_INTERFACE_CHANGE_NOTIFICATION NotificationInfos;
+                UNICODE_STRING SymbolicLinkU;
+                PWSTR SymbolicLink;
+
+                Status = IoGetDeviceInterfaces((LPGUID)EventCategoryData,
+                                               NULL, /* PhysicalDeviceObject OPTIONAL */
+                                               0, /* Flags */
+                                               &SymbolicLinkList);
+                if (NT_SUCCESS(Status))
+                {
+                    /* Enumerate SymbolicLinkList */
+                    NotificationInfos.Version = 1;
+                    NotificationInfos.Size = sizeof(DEVICE_INTERFACE_CHANGE_NOTIFICATION);
+                    RtlCopyMemory(&NotificationInfos.Event,
+                                  &GUID_DEVICE_INTERFACE_ARRIVAL,
+                                  sizeof(GUID));
+                    RtlCopyMemory(&NotificationInfos.InterfaceClassGuid,
+                                  EventCategoryData,
+                                  sizeof(GUID));
+                    NotificationInfos.SymbolicLinkName = &SymbolicLinkU;
+
+                    for (SymbolicLink = SymbolicLinkList;
+                         *SymbolicLink;
+                         SymbolicLink += (SymbolicLinkU.Length / sizeof(WCHAR)) + 1)
+                    {
+                        RtlInitUnicodeString(&SymbolicLinkU, SymbolicLink);
+                        DPRINT("Calling callback routine for %S\n", SymbolicLink);
+                        (*CallbackRoutine)(&NotificationInfos, Context);
+                    }
+
+                    ExFreePool(SymbolicLinkList);
```

Comment:
Can we use a ExFreePoolWithTag here? Or is the tag actually not known?

Reply:
Does it make sense to use `ExFreePoolWithTag(ptr, 0)` instead of `ExFreePool(ptr)`?